pixeltable/env.py

```diff
@@ -816,6 +818,91 @@ def __init_spacy(self) -> None:
         except Exception as exc:
             raise excs.Error(f'Failed to load spaCy model: {spacy_model}') from exc
 
+    def _cleanup(self) -> None:
+        """
+        Internal cleanup method that properly closes all resources and resets state.
+        This is called before destroying the singleton instance.
+        """
+        # Close database connections and sessions
+        if self._current_session is not None:
```

Comment:
Have you checked that all these conditions actually have a chance of being met? I'd expect that `_cleanup()` would only ever be called in a context where `_current_session` and `current_conn` must be `None`. Maybe these should be assertions instead? (I could be wrong about this, just not seeing it right now, but I think we should at least be able to articulate the conditions under which they might be non-`None`)

Reply:
for single threaded yes, _cleanup() gets called only when worker is moving to another db and _current_session and current_conn will be None
We can move them to assertion but I'm thinking of not raising exceptions for assertions in cloud env.